runtime/compiler/p/codegen/J9PPCWatchedStaticFieldSnippet.cpp

```diff
@@ -0,0 +1,129 @@
+    /*
+    typedef struct J9JITWatchedStaticFieldData {
+            J9Method *method;               // Currently executing method
+            UDATA location;                 // Bytecode PC index
+            void *fieldAddress;             // Address of static field storage
+            J9Class *fieldClass;            // Declaring class of static field
+    } J9JITWatchedStaticFieldData;
+    */
+
+    // Emit each field and add a relocation record (for AOT compiles) for any field if needed.
+    J9JITWatchedStaticFieldData *str = reinterpret_cast<J9JITWatchedStaticFieldData *>(cursor);
+    str->method = TR::J9WatchedStaticFieldSnippet::staticFieldData.method;
+    str->location = TR::J9WatchedStaticFieldSnippet::staticFieldData.location;
+    str->fieldAddress = TR::J9WatchedStaticFieldSnippet::staticFieldData.fieldAddress;
+    str->fieldClass = TR::J9WatchedStaticFieldSnippet::staticFieldData.fieldClass;
+
+    if (cg()->comp()->getOption(TR_UseSymbolValidationManager))
+        {
+        cg()->addExternalRelocation(
+            new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, method), reinterpret_cast<uint8_t *>(staticFieldData.method), reinterpret_cast<uint8_t *>(TR::SymbolType::typeMethod), TR_SymbolFromManager, cg()),
+            __FILE__,
+            __LINE__,
+            node);
+        }
+    else
+        {
+        cg()->addExternalRelocation(new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, method), NULL, TR_RamMethod, cg()), __FILE__, __LINE__, node);
+        }
+
+    bool isResolved = !node->getSymbolReference()->isUnresolved();
+    // If the field is unresolved then we populate these snippet fields with the correct value at runtime (via the instructions generated by generateFillInDataBlockSequenceForUnresolvedField)
+    // and hence don't need to add relocation records here.
+    if (isResolved)
+        {
+        cg()->addExternalRelocation(
+            new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, fieldAddress), reinterpret_cast<uint8_t *>(node->getSymbolReference()), reinterpret_cast<uint8_t *>(node->getInlinedSiteIndex()), TR_DataAddress, cg()),
+            __FILE__,
+            __LINE__,
+            node);
+        if (cg()->comp()->getOption(TR_UseSymbolValidationManager))
+            {
+            cg()->addExternalRelocation(
+                new (cg()->trHeapMemory()) TR::ExternalRelocation(cursor + offsetof(J9JITWatchedStaticFieldData, fieldClass), reinterpret_cast<uint8_t *>(staticFieldData.fieldClass), reinterpret_cast<uint8_t *>(TR::SymbolType::typeClass), TR_SymbolFromManager, cg()),
+                __FILE__,
+                __LINE__,
+                node);
+            }
+        else
+            {
+            /* Comment in when AOT is Enabled
```

Comment:
Would you be able to explain why this code is commented out (either here or as a comment in the codebase)? The forward path for the if doesn't comment out the code while the fallback does. So it would be good if we have extra information on why it's disabled?

Reply:
In this commit, we're not adding AOT support to Power.  
I will have another commit shortly after this to enable contiguous support for TR_ClassAddress. Due to the time squeeze, there's no certainty it will be included in R0.15. Once support is enabled, this will be uncommented. 

Either way, OpenJ9 does not have the capabilities to have AOT and FW running at the same time.

Reply:
Yes, I agree, we will not add AOT support for FW into 0.15.